### PR TITLE
refactor: mostly refactors 

### PR DESCRIPTION
**Pull request by mpoke**
## Description

Closes: #XXXX

Review in order of commits. 

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

* [ ] Included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] Added `!` to the type prefix if the change is [state-machine breaking](https://github.com/cosmos/interchain-security/blob/main/RELEASES.md#breaking-changes)
* [ ] Confirmed this PR does not introduce changes requiring state migrations, OR migration code has been added to consumer and/or provider modules
* [ ] Targeted the correct branch (see [PR Targeting](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#pr-targeting))
* [ ] Provided a link to the relevant issue or specification
* [ ] Followed the guidelines for [building SDK modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/00-intro.md)
* [ ] Included the necessary unit and integration [tests](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#testing)
* [ ] Added a changelog entry to `CHANGELOG.md`
* [ ] Included comments for [documenting Go code](https://blog.golang.org/godoc)
* [ ] Updated the relevant documentation or specification
* [ ] Reviewed "Files changed" and left comments if necessary 
* [ ] Confirmed all CI checks have passed
* [ ] If this PR is library API breaking, bump the go.mod version string of the repo, and follow through on a new major release

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

* [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] confirmed `!` the type prefix if the change is state-machine breaking
* [ ] confirmed this PR does not introduce changes requiring state migrations, OR confirmed migration code has been added to consumer and/or provider modules
* [ ] confirmed all author checklist items have been addressed
* [ ] reviewed state machine logic
* [ ] reviewed API design and naming
* [ ] reviewed documentation is accurate
* [ ] reviewed tests and test coverage

## Summary by CodeRabbit

- **New Features**
  - Enhanced consumer chain management with new methods for preparation, launching, and stopping.
  - Introduced dynamic management of consumer reward denominations.
  - Improved message handling with better validation and error tracking for governance-related messages.

- **Bug Fixes**
  - Improved error handling across various functions to ensure robustness.

- **Documentation**
  - Updated event and attribute constants for consumer rewards to streamline processing.

- **Tests**
  - Added new test cases for consumer lifecycle management and refined existing tests for clarity.

- **Chores**
  - Removed outdated functions related to consumer management for better maintainability.

### CODE REVIEW

File: x/ccv/provider/keeper/keeper.go [hunk 1]
```diff
@@ -637,9 +637,18 @@ func (k Keeper) DeletePendingVSCPackets(ctx sdk.Context, consumerId string) {
 }
 
 // SetConsumerClientId sets the client id for the given consumer id
-func (k Keeper) SetConsumerClientId(ctx sdk.Context, consumerId, clientID string) {
+func (k Keeper) SetConsumerClientId(ctx sdk.Context, consumerId, clientId string) {
```
Comment (mpoke) [inline x/ccv/provider/keeper/keeper.go:640]: TODO: please pay attention to this change

File: x/ccv/provider/keeper/distribution.go [hunk 2]
```diff
@@ -361,3 +361,29 @@ func (k Keeper) HandleSetConsumerCommissionRate(ctx sdk.Context, consumerId stri
 		commissionRate,
 	)
 }
+
+// TODO: this method needs to be tested
```
Comment (insumity) [inline x/ccv/provider/keeper/distribution.go:365]: Can we add the tests?

File: x/ccv/provider/keeper/distribution.go [hunk 3]
```diff
@@ -361,3 +361,29 @@ func (k Keeper) HandleSetConsumerCommissionRate(ctx sdk.Context, consumerId stri
...
+func (k Keeper) ChangeRewardDenoms(ctx sdk.Context, denomsToAdd, denomsToRemove []string) []sdk.Attribute {
+	eventAttributes := []sdk.Attribute{}
+	for _, denomToAdd := range denomsToAdd {
+		// Log error and move on if one of the denoms is already registered
+		if k.ConsumerRewardDenomExists(ctx, denomToAdd) {
+			ctx.Logger().Error("denom %s already registered", denomToAdd)
```
Comment (insumity) [inline x/ccv/provider/keeper/distribution.go:371]: I do not believe that format specifiers work with `ctx.Logger().Error()`. We probably need something like:
`.Error("denom already registered", "denom", denomToAdd)`.

File: x/ccv/provider/types/errors.go [hunk 4]
```diff
@@ -49,4 +49,5 @@ var (
 	ErrCannotCreateTopNChain                   = errorsmod.Register(ModuleName, 41, "cannot create Top N chain outside permissionlessly")
 	ErrCannotPrepareForLaunch                  = errorsmod.Register(ModuleName, 42, "cannot prepare chain for launch")
 	ErrInvalidStopTime                         = errorsmod.Register(ModuleName, 43, "invalid stop time")
+	ErrInvalidChangeRewardDenoms               = errorsmod.Register(ModuleName, 44, "invalid change reward denoms message")
```
Comment (insumity) [inline x/ccv/provider/types/errors.go:52]: Is `ErrInvalidChangeRewardDenoms` used anywhere? I see in a previous [commit](https://github.com/cosmos/interchain-security/pull/2194/commits/452fe04bfc724539f34ebcb9e9b3e99b467a1487) you had:
```
// ValidateBasic is not called automatically for messages from gov proposals
if err := msg.ValidateBasic(); err != nil {
    return nil, errorsmod.Wrapf(types.ErrInvalidChangeRewardDenoms, "error: %s", err.Error())
}
```
But it was then subsequently removed. Do we need to call `ValidateBasic` for messes that stem from gov proposals or not? If we do, we should do this for `CreateConsumer` and `UpdateConsumer` as well.

File: x/ccv/provider/keeper/keeper.go (hunk 1 above)
Comment (insumity) [inline x/ccv/provider/keeper/keeper.go:640]: Looks correct, albeit comments of those functions would need to be updated. However, it seems this change slightly complicates the code. What exactly is the benefit of storing the reverse at the same time?
Do we perform a similar type of reverse storing in any other `Set...` in our code? 
For example, we also have `consumerId <=> channelId` association, should it we do it there as well?

File: x/ccv/provider/keeper/distribution.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mpoke) [reply x/ccv/provider/keeper/distribution.go:365]: Yes, generate the unit testing code.

File: x/ccv/provider/types/errors.go (hunk 4 above)
Comment (mpoke) [reply x/ccv/provider/types/errors.go:52]: No, the gov module calls ValidateBasic for all messages when the proposal is submitted. I will remove the error.

File: x/ccv/provider/keeper/keeper.go (hunk 1 above)
Comment (insumity) [reply x/ccv/provider/keeper/keeper.go:640]: That's a really good point. Maybe we can create an issue to do the remaining ones.
Note, that comments need to be updated. The setter only talks about the `consumerId` to `clientId` association. Might make sense to rename the methods as well so it's clear they do the reverse store as well.

Comment (mpoke) [reply x/ccv/provider/keeper/keeper.go:642]: See https://github.com/cosmos/interchain-security/pull/2194/commits/5240deb6afbde65f72c4ec0683fc08c6c8ec737e

**Review by insumity — APPROVED** [review]

**Pull request merged**